File: src/pipeline.py
from tqdm import tqdm

from .blend_images import blend_images
from .detect_points import detect_points
from .estimate_homography import estimate_homography
from .load_images import load_images, save_image
from .match_points import match_points
from .warp_images import full_warp


def align_pair(base, image):
    keypoints_base, descriptors_base = detect_points(base)
    keypoints_image, descriptors_image = detect_points(image)
    matches = match_points(descriptors_base, descriptors_image)
    homography, inliers = estimate_homography(keypoints_base, keypoints_image, matches)
    return homography, matches, inliers


def merge_pair(base, image):
    homography, matches, inliers = align_pair(base, image)
    warped_base, warped_image, base_mask, image_mask = full_warp(base, image, homography)
    panorama = blend_images(warped_base, warped_image, base_mask, image_mask)
    info = {
        "matches": len(matches),
        "inliers": int(inliers.sum()),
        "inlier_ratio": float(inliers.mean()),
    }
    return panorama, info


def stitch_images(images):
    if len(images) < 2:
        raise ValueError("il faut au moins deux images")
    panorama = images[0]
    infos = []
    for image in tqdm(images[1:], desc="assemblage"):
        panorama, info = merge_pair(panorama, image)
        infos.append(info)
    return panorama, infos


def run_pipeline(paths, output_path, max_width=1600):
    images = load_images(paths, max_width=max_width)
    panorama, infos = stitch_images(images)
    save_image(output_path, panorama)
    return panorama, infos

File: src/pipeline.cpp
#include "pipeline.hh"

#include <opencv2/calib3d.hpp>
#include <stdexcept>

namespace panorama
{

    namespace
    {
        std::vector<cv::Mat>
        buildAbsoluteHomographies(const std::vector<cv::Mat>& pairwiseH,
                                  int ref)
        {
            int n = static_cast<int>(pairwiseH.size()) + 1;
            std::vector<cv::Mat> absH(n);
            absH[ref] = cv::Mat::eye(3, 3, CV_64F);

            for (int i = ref - 1; i >= 0; --i)
                absH[i] = absH[i + 1] * pairwiseH[i].inv();

            for (int i = ref + 1; i < n; ++i)
                absH[i] = absH[i - 1] * pairwiseH[i - 1];

            return absH;
        }
    } // namespace

    PipelineResult runPipeline(const std::vector<fs::path>& inputPaths,
                               const fs::path& outputPath)
    {
        std::vector<cv::Mat> images = loadImages(inputPaths);
        const int n = static_cast<int>(images.size());
        if (n < 2)
            throw std::runtime_error(
                "runPipeline: il faut au moins deux images");

        // features de chaque image, calculees une seule fois
        std::vector<Features> features;
        features.reserve(n);
        for (const auto& img : images)
            features.push_back(detectFeatures(img));

        cv::Mat panorama;

        if (n == 2)
        {
            auto matches = matchFeatures(features[0], features[1]);
            cv::Mat H = estimateHomography(features[0], features[1], matches);
            cv::Mat identity = cv::Mat::eye(3, 3, CV_64F);

            auto warped = warpImages({ images[0], images[1] }, { identity, H });
            panorama = blendImages(warped);
        }
        else
        {
            std::vector<cv::Mat> pairwiseH;
            pairwiseH.reserve(n - 1);
            for (int i = 0; i < n - 1; ++i)
            {
                auto matches = matchFeatures(features[i], features[i + 1]);
                pairwiseH.push_back(
                    estimateHomography(features[i], features[i + 1], matches));
            }

            int ref = n / 2;
            auto absH = buildAbsoluteHomographies(pairwiseH, ref);
            auto warped = warpImages(images, absH);
            panorama = blendImages(warped);
        }

        saveImage(outputPath, panorama);

        PipelineResult result;
        result.panorama = panorama;
        result.imagesCount = n;
        return result;
    }

} // namespace panorama